crates/factor-outbound-pg/src/client.rs

```diff
@@ -1,20 +1,89 @@
-use anyhow::{anyhow, Result};
+use anyhow::{anyhow, Context, Result};
 use native_tls::TlsConnector;
 use postgres_native_tls::MakeTlsConnector;
 use spin_world::async_trait;
 use spin_world::spin::postgres::postgres::{
     self as v3, Column, DbDataType, DbValue, ParameterValue, RowSet,
 };
 use tokio_postgres::types::Type;
-use tokio_postgres::{config::SslMode, types::ToSql, Row};
-use tokio_postgres::{Client as TokioClient, NoTls, Socket};
+use tokio_postgres::{config::SslMode, types::ToSql, NoTls, Row};
 
+/// Max connections in a given address' connection pool
+const CONNECTION_POOL_SIZE: usize = 64;
+/// Max addresses for which to keep pools in cache.
+const CONNECTION_POOL_CACHE_CAPACITY: u64 = 1024;
+
+/// A factory object for Postgres clients. This abstracts
+/// details of client creation such as pooling.
 #[async_trait]
-pub trait Client {
-    async fn build_client(address: &str) -> Result<Self>
-    where
-        Self: Sized;
+pub trait ClientFactory: Default + Send + Sync + 'static {
+    /// The type of client produced by `get_client`.
+    type Client: Client;
+    /// Gets a client from the factory.
+    async fn get_client(&self, address: &str) -> Result<Self::Client>;
+}
+
+/// A `ClientFactory` that uses a connection pool per address.
+pub struct PooledTokioClientFactory {
+    pools: moka::sync::Cache<String, deadpool_postgres::Pool>,
+}
+
+impl Default for PooledTokioClientFactory {
+    fn default() -> Self {
+        Self {
+            pools: moka::sync::Cache::new(CONNECTION_POOL_CACHE_CAPACITY),
```

Comment:
@lann I'd value your input on whether we need to provide a way for hosts (e.g. multi-tenant environments) to configure cache (and pool) sizes and policies (e.g. idle time eviction).  But also happy to punt that to a separate PR if that's not yet obvious.